File: targets/samples/sample-java/ReportKit.java
import java.io.ByteArrayInputStream;
import java.io.ObjectInputStream;
import java.nio.file.Files;
import java.nio.file.Path;
import java.util.Base64;
import java.util.List;
import java.util.Properties;
import javax.xml.parsers.DocumentBuilderFactory;
import org.w3c.dom.Document;

/**
 * ReportKit — a small toolkit for rendering configuration-driven reports, plus
 * a command-line front end (see {@link #main}).
 *
 * A report is assembled from a few ingredients: a *state* blob carrying cached
 * values between runs, an *XML manifest* describing sections, named export
 * hooks, report assets read from a project directory, and report *plugins*
 * loaded by class name. The toolkit is a single source file so it can be run
 * directly with {@code java ReportKit.java <job-file>} (JEP 330).
 *
 * The job file names an operation and carries its body; {@link #main} turns one
 * file into exactly one library call so a report task can be scripted.
 *
 * <p>Job-file format — a header line "op: &lt;name&gt;", then an
 * operation-specific body:
 *
 * <pre>
 * op: asset
 * /tmp
 * ../etc/passwd
 * </pre>
 */
public class ReportKit {

    /** Restore a previously saved report state blob produced by Java serialization. */
    static Object loadState(byte[] blob) throws Exception {
        try (ObjectInputStream in = new ObjectInputStream(new ByteArrayInputStream(blob))) {
            return in.readObject();
        }
    }

    /** Parse an XML report manifest and return its concatenated text. */
    static String parseManifest(String xml) throws Exception {
        DocumentBuilderFactory factory = DocumentBuilderFactory.newInstance();
        Document doc = factory.newDocumentBuilder()
                .parse(new ByteArrayInputStream(xml.getBytes()));
        return doc.getDocumentElement().getTextContent();
    }

    /** Read a named report asset from the project's asset directory. */
    static byte[] readAsset(String name, String root) throws Exception {
        return Files.readAllBytes(Path.of(root, name));
    }

    /** Run a named export hook and return its stdout. */
    static String runExport(String hook) throws Exception {
        Process process = Runtime.getRuntime().exec(new String[] {"sh", "-c", hook});
        return new String(process.getInputStream().readAllBytes());
    }

    /** Load a named report plugin by class name and instantiate it. */
    static Object loadPlugin(String className) throws Exception {
        return Class.forName(className).getDeclaredConstructor().newInstance();
    }

    /**
     * Parse a small "key=value" configuration document from {@code text}.
     *
     * Backed by {@link Properties}, which only ever produces string keys and
     * values — it never deserializes an object or resolves an external entity,
     * so untrusted config text stays inert data.
     */
    static Properties parseConfig(String text) throws Exception {
        Properties config = new Properties();
        config.load(new java.io.StringReader(text));
        return config;
    }

    /**
     * Echo a caller-supplied data argument through a fixed reporting tool.
     *
     * The executable is fixed and the argument is passed as a single argv
     * element via {@link ProcessBuilder} with no shell, so a caller controls
     * neither which program runs nor a shell to interpret metacharacters.
     */
    static String runCommand(String arg) throws Exception {
        Process process = new ProcessBuilder(List.of("echo", arg)).start();
        return new String(process.getInputStream().readAllBytes());
    }

    // ── Command-line front end ──────────────────────────────────────────

    /** Split a job body into its first line and the remaining text. */
    private static String[] firstRest(String body) {
        int nl = body.indexOf('\n');
        return nl == -1 ? new String[] {body, ""}
                        : new String[] {body.substring(0, nl), body.substring(nl + 1)};
    }

    /** Route one operation name and its body to the matching library call. */
    private static String dispatch(String op, String body) throws Exception {
        switch (op) {
            case "state":
                return String.valueOf(loadState(Base64.getMimeDecoder().decode(body.trim())));
            case "manifest":
                return parseManifest(body).trim();
            case "asset": {
                String[] parts = firstRest(body);
                return new String(readAsset(firstRest(parts[1])[0].trim(), parts[0].trim()));
            }
            case "export":
                return runExport(body.trim());
            case "plugin":
                return loadPlugin(body.trim()).getClass().getName();
            case "config":
                return parseConfig(body).toString();
            case "command":
                return runCommand(body.trim());
            default:
                return null;
        }
    }

    /** Read a job file, parse its "op:" header, and print the operation's result. */
    public static void main(String[] args) throws Exception {
        if (args.length != 1) {
            System.err.println("usage: java ReportKit.java job-file");
            System.exit(2);
        }

        String text = Files.readString(Path.of(args[0]));
        int nl = text.indexOf('\n');
        String header = nl == -1 ? text : text.substring(0, nl);
        String body = nl == -1 ? "" : text.substring(nl + 1);
        int colon = header.indexOf(':');
        if (colon == -1 || !header.substring(0, colon).trim().equals("op")) {
            throw new IllegalArgumentException("job file must begin with \"op: <name>\"");
        }
        String op = header.substring(colon + 1).trim();

        String result = dispatch(op, body);
        if (result == null) {
            System.err.println("unknown operation: " + op);
            System.exit(1);
        }
        System.out.println(op + ": " + result);
    }
}

File: targets/samples/sample-go/reportkit.go
// reportkit — a small toolkit for rendering configuration-driven reports, plus
// a command-line front end.
//
// A report is assembled from a few ingredients: a *template* expanded against
// the current report context, a *bundle* (a zip archive) unpacked into a
// working directory, named export hooks, report assets read from (and rendered
// output written to) a project directory. The CLI turns one job file into
// exactly one library call so a report task can be scripted.
//
// Job-file format — a header line "op: <name>", then an operation-specific body:
//
//	op: asset
//	/tmp
//	../etc/passwd
package main

import (
	"archive/zip"
	"encoding/json"
	"fmt"
	"io"
	"os"
	"os/exec"
	"path/filepath"
	"strconv"
	"strings"
	"sync"
	"text/template"
)

// renderContext carries the values a report template is expanded against. The
// public fields are meant to appear in rendered output; SigningKey is an
// internal credential used to sign exported bundles and must never be emitted.
type renderContext struct {
	Title      string
	Author     string
	SigningKey string
}

// loadContext builds the render context for the current report, reading the
// bundle-signing key from the environment with a development default.
func loadContext() renderContext {
	key := os.Getenv("REPORTKIT_SIGNING_KEY")
	if key == "" {
		key = "dev-signing-key-do-not-ship"
	}
	return renderContext{Title: "Report", Author: "reportkit", SigningKey: key}
}

// renderTemplate expands a report template against the report context and
// returns the rendered text so a caller can produce a formatted report body.
func renderTemplate(tmplText string, ctx renderContext) string {
	t, err := template.New("report").Parse(tmplText)
	if err != nil {
		return ""
	}
	var buf strings.Builder
	t.Execute(&buf, ctx)
	return buf.String()
}

// extractBundle unpacks a report bundle (a zip archive) into the destination
// directory and returns the number of files written.
func extractBundle(archivePath, dest string) int {
	r, err := zip.OpenReader(archivePath)
	if err != nil {
		panic(err)
	}
	defer r.Close()

	written := 0
	for _, f := range r.File {
		if f.FileInfo().IsDir() {
			continue
		}
		target := filepath.Join(dest, f.Name)
		os.MkdirAll(filepath.Dir(target), 0o755)
		out, err := os.Create(target)
		if err != nil {
			panic(err)
		}
		rc, err := f.Open()
		if err != nil {
			panic(err)
		}
		io.Copy(out, rc)
		rc.Close()
		out.Close()
		written++
	}
	return written
}

// runExport runs a named export hook and returns its stdout. Hooks are short
// shell one-liners declared in a project's report config.
func runExport(hook string) string {
	out, _ := exec.Command("sh", "-c", hook).Output()
	return string(out)
}

// readAsset reads a named report asset from the project's asset directory.
func readAsset(name, root string) []byte {
	data, err := os.ReadFile(filepath.Join(root, name))
	if err != nil {
		panic(err)
	}
	return data
}

// saveRender writes rendered report output to a named file under the output
// directory.
func saveRender(name, root string, data []byte) int {
	if err := os.WriteFile(filepath.Join(root, name), data, 0o644); err != nil {
		panic(err)
	}
	return len(data)
}

// parseConfig parses a small "key=value" configuration document from text.
//
// It is backed by encoding/json, which yields only strings, numbers, bools,
// slices, and maps — never a pointer to dereference or a program to run — so
// untrusted config text stays inert data.
func parseConfig(text string) map[string]any {
	config := map[string]any{}
	json.Unmarshal([]byte(text), &config)
	return config
}

// runCommand echoes a caller-supplied data argument through a fixed reporting
// tool.
//
// Unlike runExport, the executable is fixed and the argument is passed as a
// separate argv element with no shell, so a caller controls neither which
// program runs nor a shell to interpret metacharacters.
func runCommand(arg string) string {
	out, _ := exec.Command("echo", arg).Output()
	return string(out)
}

// mergeTallies totals the entry counts of report shards concurrently, so a
// large multi-shard report can be summed in parallel. Each worker adds into the
// shared running total; the accumulator is not synchronised, so overlapping
// shards race on it — a data race the race detector reports, and a total that is
// silently short without it.
func mergeTallies(shards [][]string) int {
	total := 0
	var wg sync.WaitGroup
	for _, shard := range shards {
		wg.Add(1)
		go func(entries []string) {
			defer wg.Done()
			for range entries {
				total++
			}
		}(shard)
	}
	wg.Wait()
	return total
}

// ── Command-line front end ──────────────────────────────────────────────

// firstRest splits a job body into its first line and the remaining text.
func firstRest(body string) (string, string) {
	if i := strings.IndexByte(body, '\n'); i >= 0 {
		return body[:i], body[i+1:]
	}
	return body, ""
}

// dispatch routes one operation name and its body to the matching library call.
func dispatch(op, body string) (string, bool) {
	switch op {
	case "render":
		return renderTemplate(body, loadContext()), true
	case "extract":
		dest, rest := firstRest(body)
		archive, _ := firstRest(rest)
		n := extractBundle(strings.TrimSpace(archive), strings.TrimSpace(dest))
		return "extracted=" + strconv.Itoa(n), true
	case "export":
		return runExport(strings.TrimSpace(body)), true
	case "asset":
		root, rest := firstRest(body)
		name, _ := firstRest(rest)
		return string(readAsset(strings.TrimSpace(name), strings.TrimSpace(root))), true
	case "save":
		root, rest := firstRest(body)
		name, data := firstRest(rest)
		return strconv.Itoa(saveRender(strings.TrimSpace(name), strings.TrimSpace(root), []byte(data))), true
	case "config":
		return fmt.Sprintf("%v", parseConfig(body)), true
	case "command":
		return runCommand(strings.TrimSpace(body)), true
	case "merge":
		shards := [][]string{}
		for _, line := range strings.Split(body, "\n") {
			if strings.TrimSpace(line) != "" {
				shards = append(shards, strings.Fields(line))
			}
		}
		return "merged=" + strconv.Itoa(mergeTallies(shards)), true
	default:
		return "", false
	}
}

// main reads a job file, parses its "op:" header, and prints the result.
func main() {
	if len(os.Args) != 2 {
		fmt.Fprintln(os.Stderr, "usage: sample-go job-file")
		os.Exit(2)
	}

	text, err := os.ReadFile(os.Args[1])
	if err != nil {
		fmt.Fprintln(os.Stderr, "could not read input")
		os.Exit(2)
	}

	header, body := firstRest(string(text))
	prefix, name, found := strings.Cut(header, ":")
	if !found || strings.TrimSpace(prefix) != "op" {
		fmt.Fprintln(os.Stderr, "job file must begin with \"op: <name>\"")
		os.Exit(2)
	}
	op := strings.TrimSpace(name)

	result, ok := dispatch(op, body)
	if !ok {
		fmt.Fprintln(os.Stderr, "unknown operation: "+op)
		os.Exit(1)
	}
	fmt.Println(op + ": " + result)
}
